miniwin/miniwin/src/miniwin_ddraw.cpp

```diff
@@ -171,21 +171,54 @@ HRESULT DirectDrawImpl::EnumDisplayModes(
 	LPDDENUMMODESCALLBACK lpEnumModesCallback
 )
 {
-	DDSURFACEDESC ddsd = {};
-	ddsd.dwSize = sizeof(DDSURFACEDESC);
-	ddsd.dwFlags = DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT;
-	ddsd.dwWidth = 640;
-	ddsd.dwHeight = 480;
-	ddsd.lPitch = 0;
-
-	ddsd.ddpfPixelFormat.dwSize = sizeof(DDPIXELFORMAT);
-	ddsd.ddpfPixelFormat.dwFlags = DDPF_RGB | D3DDD_DEVICEZBUFFERBITDEPTH;
-	ddsd.ddpfPixelFormat.dwRGBBitCount = 16; // Game only accpets 8 or 16
-
-	if (!lpEnumModesCallback(&ddsd, lpContext)) {
+	if (!lpEnumModesCallback) {
 		return DDERR_GENERIC;
 	}
 
+	SDL_DisplayID displayID = SDL_GetPrimaryDisplay();
+	if (!displayID) {
+		return DDERR_GENERIC;
+	}
+
+	int numModes = 0;
+	SDL_DisplayMode** modes = SDL_GetFullscreenDisplayModes(displayID, &numModes);
+	if (numModes <= 0) {
+		return DDERR_GENERIC;
+	}
+
+	for (int i = 0; i < numModes; ++i) {
+		const SDL_DisplayMode* mode = modes[i];
+
+		SDL_Surface* surface = SDL_CreateSurface(mode->w, mode->h, mode->format);
+		if (!surface) {
+			continue;
+		}
+
+		DDSURFACEDESC ddsd = {};
+		ddsd.dwSize = sizeof(DDSURFACEDESC);
+		ddsd.dwFlags = DDSD_WIDTH | DDSD_HEIGHT;
+		ddsd.dwWidth = surface->w;
+		ddsd.dwHeight = surface->h;
+		ddsd.lPitch = surface->pitch;
+
+		const SDL_PixelFormatDetails* format = SDL_GetPixelFormatDetails(surface->format);
```

Comment:
`SDL_Surface` is only RAM memory: it does not touch the video.
```suggestion
		const SDL_DisplayMode* mode = modes[i];
		const SDL_PixelFormatDetails* format = SDL_GetPixelFormatDetails(mode->format);

		DDSURFACEDESC ddsd = {};
		ddsd.dwSize = sizeof(DDSURFACEDESC);
		ddsd.dwFlags = DDSD_WIDTH | DDSD_HEIGHT;
		ddsd.dwWidth = mode->w;
		ddsd.dwHeight = mode->h;
```

[This documentation](https://learn.microsoft.com/en-us/windows/win32/direct3ddds/dds-header) says `lPitch` is only set when dwFlags contains `DDSD_PITCH`.

Reply:
Yeah, it's a bit hard to follow though since Isle doesn't ever checks the flag and just assumes pitch is there when it needs it. But on closer inspection it only does so after calling `DirectDrawSurface::Lock()` so your right it's not needed here.